feat: added filter functionality.

MillenniumFalconMechanic opened the pull request:
### Reviewers
**Functional:** 
@tihuan 

---

## Changes
Added filter functionality behind `filter` feature flag.

Included in this PR are:
* [#1613 Implement core facet search module/engine](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1613).
* [#1609 Read datasets and collections on load](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1609).
* [#1602 Update collections UI to match mocks](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1602).
* [#1614 List basic metadata categories in left sidebar on collections view](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1614).
* [#1604 Add datasets view](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1604).
* [#1605 Update datasets table on collection detail page](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1605).

## Notes
#### [#1605 Update datasets table on collection detail page](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1605)
* The updates to the collection datasets table only includes UI changes and does not include a conversion to react-table.
* To reduce changes to the existing collections datasets table (outside of the filter flag), we have duplicated the component hierarchy and styled the duplicated components. 
* Existing structure can be removed once collection datasets table is converted to react-table but this requires a review of the dataset fetch model and status checks.

## Definition of Done (from ticket)
* See individual tickets.

## QA steps (optional)
* Datasets located at `/datasets`.
* Collections located at `/collections`.
* Datasets table on collection detail page *outside of feature flag* should be reviewed.

## Known Issues
* Dataset downloads are not currently hooked up and will be resolved by [#1724 Use dataset-metadata endpoint to populate dataset download modal.](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1724)
* Explorer URL is hooked up but won't be end-to-end until [#1721 Add explorer_url to datasets/index](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1721) is merged.
* Redirect from `/ `to `/datasets` is not currently hooked up and will be resolved by [1718 Remove filter feature flag](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1718).

MillenniumFalconMechanic commented on frontend/src/common/entities.ts:
```diff
@@ -53,9 +53,9 @@ export enum X_APPROXIMATE_DISTRIBUTION {
-  PRIMARY = "primary",
-  SECONDARY = "secondary",
-  BOTH = "both",
+  PRIMARY = "PRIMARY",
```
Uppercase matches endpoint response values.

MillenniumFalconMechanic commented on frontend/src/components/Collections/components/Grid/components/Row/DatasetRow/index.tsx:
```diff
@@ -265,11 +265,4 @@ const DatasetRow: FC<Props> = ({
   );
 };
 
-/** Maximum number of cells a dataset can have in order to be included for display. */
```
Moved to utils to enabled reuse.

MillenniumFalconMechanic commented on frontend/src/components/Header/index.tsx:
```diff
@@ -15,29 +16,54 @@ import {
   const { data: userInfo } = useUserInfo(isCurator);
-  const { asPath } = useRouter();
-  const isMyCollectionsActive = asPath === ROUTES.MY_COLLECTIONS;
+  const { pathname } = useRouter();
```
Using `pathname` rather than `asPath` to exclude query string parameters.

tihuan commented on frontend/src/components/common/Filter/components/FilterMenu/index.tsx:
```diff
@@ -0,0 +1,104 @@
+  categoryKey: CATEGORY_KEY;
+  filterCategoryValues: FilterCategoryValuesFn;
+  filterCategoryValuesWithCount: FilterCategoryValuesWithCountFn;
+  multiselect: boolean;
```
nit: `isMultiSelect`

MillenniumFalconMechanic replied:
[Updated](https://github.com/chanzuckerberg/single-cell-data-portal/blob/cc/filter/frontend/src/components/common/Filter/components/FilterMenu/index.tsx#L25)!

tihuan commented on frontend/src/components/common/Filter/components/FilterMenu/index.tsx:
```diff
@@ -0,0 +1,104 @@
+  multiselect: boolean;
+  onFilter: OnFilterFn;
+  onUpdateSearchValue: OnUpdateSearchValueFn;
+  searchable: boolean;
```
nit: `isSearchable`

MillenniumFalconMechanic replied:
[Updated](https://github.com/chanzuckerberg/single-cell-data-portal/blob/cc/filter/frontend/src/components/common/Filter/components/FilterMenu/index.tsx#L26)!

tihuan commented on frontend/src/components/common/Modal/index.tsx:
```diff
@@ -2,14 +2,21 @@ import { FC } from "react";
+      isOpen={isOpen}
+      onClose={onClose}
+      className={className}
+      shouldReturnFocusOnClose={false}
```
Ooh awesome solution for fixing the tooltip issue 🤩 Thank you for figuring this out!

```suggestion
      // Remove this prop after upgrading Blueprint Tooltip library
      // Current bug is after closing the modal, tooltip will reappear, as shown in the link below:
      // https://codesandbox.io/s/blueprint-tooltip-bug-forked-jw9z2?file=/src/index.tsx
      shouldReturnFocusOnClose={false}
```

tihuan commented on frontend/src/types/react-table-config.d.ts:
```diff
@@ -0,0 +1,44 @@
+/* eslint-disable @typescript-eslint/no-empty-interface */
```
Thanks so much for adding the type definitions 🤩 🙏 🙏 🙏

tihuan reviewed (approved):
Really high quality PR 🙏 ✨ I learned a lot from reading through this! Thank you SO much for the amazing work, team 🤩 👏 🚀 !

Hopefully the memoization investigation is successful 🤞 If there's anything I can assist please let me know too!

Thank you!

MillenniumFalconMechanic commented on frontend/src/common/hooks/useCategoryFilter.ts:
```diff
@@ -0,0 +1,529 @@
+  const [filterState, setFilterState] = useState<FilterState>();
+
+  // Set up original, full set of categories and their values.
+  useMemo(() => {
```
Thanks Timmy. I think my understanding of `useMemo` might be off; I'll review the docs again.

MillenniumFalconMechanic commented:
@tihuan, thank you again for taking the time to review this PR; I have updated the code as per your review comments.

I have added [added one fix](https://github.com/chanzuckerberg/single-cell-data-portal/pull/1726/commits/dc2357f2ee0bef91f5ea7e4ef40f8b4344f4da89) to correct the display of the selected tag labels.

tihuan reviewed (approved):
ahh great catch thanks so much, @MillenniumFalconMechanic 👍 🥇 !

tihuan reviewed (approved):
Whew much cleaner dependency 😆 Thanks for updating the code, @MillenniumFalconMechanic 🙏 !!
